proposed/0016-launcher-data_types-api/0016-launcher-data-types-api.md

```diff
@@ -0,0 +1,282 @@
+i.e within the permissible range 10,001 to 2^64
+
+##### Request
+```javascript
+{
+  endpoint: 'safe-api/v1.0/data_types/create_structured_data',
+  data: {
+    id: String, // Id of the structured Data
+    tag_type: Integer,
+    content: String, // data to be saved
+    is_private: Boolean // If private the data would be encrypted
+  }
+}
+```
+##### Response
+ Returns error with code and description. If the request is processed successfully
+ then the error is returned as null
+
+###### On Success
+```javascript
+{
+  id: String, // base64 String
+  error: null
+}
+```
+
+###### On Error
+```javascript
+{
+  id: String, // base64 String
+  error: {
+    code: -200,
+    description: 'Some Error'
+  }
+}
+```
+
+#### Get structured data
+##### Request
+```javascript
+{
+  endpoint: 'safe-api/v1.0/data_types/get_structured_data',
+  data: {
+    id: String, // base64 String
+    tag_type: Integer,
+    is_private: Boolean
+  }
+}
+```
+
+##### Response
+
+###### On Success
+```javascript
+{
+  id: String, // base64 String
+  data: {
+    version: Integer, // version number of the Structured Data
+    owners: [ String ], // List of base64 String representing the public signing key of the Owners
+    content: String // base64 String
+  }
+}
+```
+
+###### On Error
+```javascript
+{
+  id: String,// base64 String
+  error: {
+    code: -200,
+    description: 'Some Error'
+  }
+}
+```
+
+#### Update structured data
+```javascript
+{
+  endpoint: 'safe-api/v1.0/data_types/update_structured_data',
+  data: {
+    id: String, // base64 String
+    tag_type: Integer,
+    version: Integer, // Incremented version number    
+    content: String, // base64 String - New Content associated to the Structured Data
+    owners: [ String ], // List of base64 String representing the public signing key of the Owners
+    previous_owners: [ String ], // Optional field - List of base64 String representing the public signing key of the previous owners.
+                                 // This is used while transferring the ownership of the Structured data
+    is_private: Boolean // If private the data would be encrypted and saved in the network
+  }
+}
+```
+
+##### Response
+ Returns error with code and description. If the request is processed successfully
+ then the error is returned as null
+
+
+###### On Success
+```javascript
+{
+ id: String, // base64 String
+ error: null
+}
+```
+
+###### On Error
+```javascript
+{
+ id: String, // base64 String
+ error: {
+   code: -200,
+   description: 'Some Error'
+ }
+}
+```
+
+## Immutable data
+
+Raw Data in the network is stored as Immutable data after the SelfEncryption process.
+
+### Save Raw data
+
+#### Request
+```javascript
+{
+  endpoint: 'safe-api/v1.0/data_types/save_raw_data',
+  data: {
+    content: String // base64 String
+  }
+}
+```
```

Comment:
One of the features discussed to make the REST API more useful was GET/PUT/POST of particular versions - these do nothing, but return an error if the version found is not the one referred to. 

This is a fairly common standard for REST/CRUD interfaces, because it makes collaborative working possible with just the REST API (I think - going from memory here).

I mention it (possibly not in the best place, sorry) so it doesn't get forgotten. I understand that there may be other planned ways to support collaborative use of storage, but this is worth considering if feasible (and it would seem to be straightforward?) because many existing applications would sit easily on this.